database.go

```diff
@@ -0,0 +1,272 @@
+package main
+
+import (
+	"sort"
+	"sync/atomic"
+)
+
+type Database interface {
+	BulkWrite([]*BulkMetric)
+	Open()
+	Close()
+	GetMedian() int
+}
+
+type MedianDatabase struct {
+	writeCh chan []*BulkMetric
+	quitCh  chan bool
+
+	// used to keep the left and right in sync!
+	left   []BulkMetric
+	right  []BulkMetric
+	size   int
+	median int32
+}
+
+func NewMedianDatabase() *MedianDatabase {
+	return &MedianDatabase{
+		writeCh: make(chan []*BulkMetric),
+		quitCh:  make(chan bool),
+		median:  0,
+	}
+}
+
+func (m *MedianDatabase) Open() {
+	go func() {
+		m.worker()
+	}()
+}
+
+func (m *MedianDatabase) Close() {
+	// signal the database to close itself
+	m.quitCh <- true
+
+	// wait for the database to finish processing
+	<-m.quitCh
+	close(m.quitCh)
+	close(m.writeCh)
+}
+
+func (m MedianDatabase) GetMedian() int {
+	// load the median value, ensuring that if this happens at the same
+	// time an update is happening that invalid data isn't accidentally
+	// returned
+	median := atomic.LoadInt32(&(m.median))
+
+	// by default int will correspond to the size of the current OS in go.
+	// Thus, it is safe to assume that this call will operate normally
+	// unless we are on a 16bit system (which is likely to cause much
+	// bigger problems elsewhere)
+	return int(median)
+}
+
+func (m *MedianDatabase) BulkWrite(bulkMetrics []*BulkMetric) {
+	// NOTE this is totally slow an unoptimized in every way possible; this
```

Comment:
In a future pull-request, let's create a type around `[]*BulkMetric` where we implement the Sort Interface https://golang.org/pkg/sort/#Interface